tools/dynamic_framework_slicer/dynamic_framework_slicer.sh

```diff
@@ -0,0 +1,54 @@
+#!/bin/bash 
+set -eu
+
+BINARIES+=()
+while [[ $# -gt 0 ]]; do
+  arg="$1"
+  case $arg in
+      --in)
+      readonly IN="$2"
+      shift
+      ;;
+      --out)
+      readonly OUT="$2"
+      shift
+      ;;
+      --realpath)
+      readonly REALPATH="$2"
+      shift
+      ;;
+      *)
+      BINARIES+=("$1")
+      ;;
+  esac
+  shift
+done
+
+# Strip out any unnecessary slices from embedded dynamic frameworks to save space
+
+for bin in $BINARIES; do
+    # Lets figure out what slices the app has
+    bin_slices=$(xcrun lipo -info "$bin" | cut -d: -f3 | sort)
+
+    if [ -f "$IN" ]; then
+        # We need to check we have overlapping architectures
+        framework_slices=$(xcrun lipo -info "$IN" | cut -d: -f3 |sort)
+        
+        # The normal case should be that the framework and app have exactly the same slices.
+        # Lets check that and skip processing if that is the case
+        if [ "$bin_slices" == "$framework_slices" ]; then
+            continue;
+        fi
+
+        lipo_args=""
+        # Now make the framework match the slices the app has
+        for slice in $bin_slices ; do
+            lipo_args="$lipo_args -extract $slice"
+        done
+        xcrun lipo "$IN" $lipo_args -output "$OUT"
+        if [ $? -ne 0 ]; then
+            echo "error: lipo of $framework_name failed" >&2
+            exit 1
+        fi
+    fi
+done
```

Comment:
How does this loop work if there are >1 BINARY to check against?

On any iteration, it will strip unneeded archs, but that isn't correct.  i.e. - 

BINARY1 - archA, archB & archC
BINARY2 - archB & archC
IN = archA, archB, & archC

Loop pass 1 -
  OUT doesn't get created because the arches match
Loop pass 2 -
  OUT gets created with archB & archC

If you make the arches matching case also copy, you still can end up doing extra work on the second loop iteration because you copy arches you already had in OUT.  The better approach is likely going to be to collect all the needed arches, then compare that to the provide ones, and then copy the file or the subset of arches once.

ps - you also need to handle the case where a framework might not have an arch a binary has (because of conditional code), i.e. - BINARY has arm7, arm64, but IN only has arm64.  Currently the script looks like it might try to extract arm7 and thus fail.

Reply:
Surely if the binary has architectures that are missing from the dynamic library, that should fail the build? Actually, the link step should fail in that scenario.

Comment about multiple binaries is good, I will fix that.

Reply:
The app might have conditional code specific to arm64 for some feature (say something AR/VR related) and thus ends up with that framework to provide the support, but arm7 build doesn't have the feature and thus doesn't need the dependency.  Hence the app can have arches the framework doesn't have.

Reply:
Hmm yes good call. Will have to rely on the linker for that scenario, which makes the script simpler. I wont try and validate the minimum set of slices, just remove the unused ones.

Reply:
@kastiglione this was the original comment thread where we got to this.